File: Space.hs

module Space where

data Pos = Pos
  { xPos :: Int
  , yPos :: Int
  } deriving (Eq, Ord, Read, Show)

data Vec = Vec
  { xDiff :: Int
  , yDiff :: Int
  } deriving (Eq, Ord, Read, Show)

distance :: Pos -> Pos -> Double
distance p1 p2 = sqrt $ fromIntegral $ toInteger $
  ((xPos p1 - xPos p2) ^ 2) + ((yPos p1 - yPos p2) ^ 2)

minus :: Pos -> Pos -> Vec
minus p1 p2 = Vec (xPos p1 - xPos p2) (yPos p1 - yPos p2)

plus :: Pos -> Vec -> Pos
plus p v = Pos (xPos p + xDiff v) (yPos p + yDiff v)

File: space.js

const Vec = Struct("Vec", [['xDiff', Int], ['yDiff', Int]], function() { });

const Pos = Struct("Pos", [['xPos', Int], ['yPos', Int]], function() {
  this.distance = typecheck("Pos.distance", [Pos], function(other) {
    return Math.sqrt(
      Math.pow(this.xPos - other.xPos, 2) + Math.pow(this.yPos - other.yPos, 2));
  });
  this.minus = typecheck("Pos.minus", [Pos], function(other) {
    return new Vec(this.xPos - other.xPos, this.yPos - other.yPos);
  });
  this.plus = typecheck("Pos.plus", [Vec], function(vec) {
    return new Pos(this.xPos + vec.xDiff, this.yPos + vec.yDiff);
  });
});

